File: src/core/protocol.ts
/**
 * The water temperature schedule the egg's surface actually sees, across every
 * phase of the cook: pan ramp, boil, and cooling.
 */

import {
  RAMP_R, TAU_AIR, T_ICE_BATH_C, T_COLD_TAP_C, T_ROOM_C,
  TAU_DIP_RECOVERY, TAU_STANDING_SCALE, C_WATER, C_EGG, TAU_PLUNGE,
} from './constants.js';

/** Cold start: eggs go in the cold pan and heat with the water. Hot start:
 *  eggs are lowered into water already boiling - peels far better, and the
 *  ramp becomes irrelevant. */
export type StartMode = 'cold' | 'hot';

/** What happens after the egg comes out. This is not a detail: it changes the
 *  peak yolk temperature by ~20 C. */
export type Cooling = 'ice' | 'tap' | 'counter';

/** What happens to the burner once the water boils. 'hold' keeps the water at
 *  its boiling point, which is what every recipe silently assumes. 'off' is the
 *  standing method - cover the pan, kill the heat, and let a falling water
 *  temperature finish the egg. */
export type HeatAfterBoil = 'hold' | 'off';

export interface CookSetup {
  startMode: StartMode;
  /** Egg temperature when it goes in, C (fridge ~4, room ~20). */
  eggStart_C: number;
  /** Room air temperature, C - the pan starts here on a cold start. */
  ambient_C: number;
  /** Boiling point at the user's altitude, C. */
  boiling_C: number;
  /** Measured time for the pan to reach a full rolling boil, s. Cold start only. */
  timeToBoil_s: number;
  cooling: Cooling;
  /** Water volume, litres. Sets how far the water dips when eggs go in, and -
   *  with the heat off - how long the pan holds its temperature. */
  waterLitres: number;
  /** Burner after the boil. Omitted means 'hold', which is what every recipe
   *  assumes without saying so. */
  afterBoil?: HeatAfterBoil;
  eggCount: number;
  eggMass_kg: number;
}

/**
 * Pan heating ramp. Constant power into a lumped water mass with Newtonian
 * losses gives an exponential approach to a steady state the pan never
 * reaches, because boiling clamps it:
 *
 *   T(t) = Tamb + r*(Tboil - Tamb)*(1 - exp(-t/tau))
 *
 * The overshoot ratio r = P/(U*dT) has a directly observable meaning: 1/r is
 * the fraction of full burner power needed to HOLD a boil, which measured
 * cooktop studies put near 1/3. One user measurement (time to boil) plus that
 * one shape constant determines the whole curve:
 *
 *   tau = t_boil / ln(r/(r-1))
 *
 * A linear ramp is the r -> infinity limit; at r = 3 the two differ by about
 * 4 C at the midpoint, worth ~30 s of cook time on a 10-minute ramp.
 */
export function rampTemperature(
  t_s: number, timeToBoil_s: number, ambient_C: number, boiling_C: number,
): number {
  if (t_s >= timeToBoil_s) return boiling_C;
  const tau = timeToBoil_s / Math.log(RAMP_R / (RAMP_R - 1.0));
  const t = ambient_C + RAMP_R * (boiling_C - ambient_C) * (1.0 - Math.exp(-t_s / tau));
  return t > boiling_C ? boiling_C : t;
}

/**
 * How far the water temperature drops when cold eggs are dropped into boiling
 * water, C. Straight energy balance over water + eggs. Four 60 g fridge eggs
 * into 2 L drops it ~8 C; into 1 L, ~14 C. Worth modelling - it is why the
 * same recipe fails in a small pan.
 */
export function dipMagnitude(setup: CookSetup): number {
  const waterCapacity = setup.waterLitres * C_WATER;
  const eggCapacity = setup.eggCount * setup.eggMass_kg * C_EGG;
  const total = waterCapacity + eggCapacity;
  if (total <= 0.0) return 0.0;
  return eggCapacity * (setup.boiling_C - setup.eggStart_C) / total;
}

/**
 * The pan's Newtonian loss time constant, seconds.
 *
 * tau = m*c/(U*A) is the same quantity that shapes the ramp, so the user's one
 * measurement - time to a rolling boil - already contains it:
 *
 *   T(t) = Tamb + r*(Tboil - Tamb)*(1 - exp(-t/tau))  reaches Tboil at
 *   t_boil = tau * ln(r/(r-1))
 *
 * It therefore scales with water volume without being told to, which is the
 * whole reason the standing method works in a stockpot and not in a milk pan.
 * TAU_STANDING_SCALE holds open the question of whether the constant is really
 * the same with the burner off and a lid on; see constants.ts.
 */
export function panTimeConstant(timeToBoil_s: number): number {
  return TAU_STANDING_SCALE * timeToBoil_s / Math.log(RAMP_R / (RAMP_R - 1.0));
}

/**
 * Water temperature once the heat is off: Newtonian cooling of the pan toward
 * the room, starting from `from_C`.
 *
 * Nothing else in the cook phase changes. The modal solver is driven by an
 * arbitrary piecewise-linear surface temperature, so a falling bath costs no
 * more than a constant one - the egg simply sees a schedule that runs out of
 * heat instead of one that does not.
 */
export function standingTemperature(
  elapsedSinceOff_s: number, from_C: number, ambient_C: number, timeToBoil_s: number,
): number {
  const tau = panTimeConstant(timeToBoil_s);
  if (!(tau > 0.0)) return from_C;
  return ambient_C + (from_C - ambient_C) * Math.exp(-elapsedSinceOff_s / tau);
}

/**
 * Surface temperature during cooling.
 *
 * Ice water and a cold tap are effectively Dirichlet - the surface goes to the
 * bath temperature and stays there. Still air is not: the egg is nearly lumped
 * (Bi ~ 0.4) and cools with a time constant of ~34 minutes, which is about 7x
 * SLOWER than it equilibrates internally (~4.5 min). So a rested egg carries
 * over almost adiabatically, and its surface tracks its own bulk temperature
 * decaying toward the room.
 *
 * `startFrom_C` is the egg's VOLUME-AVERAGE temperature at the moment it left
 * the water, not the water temperature. A lumped body relaxes from its own
 * bulk temperature; using the water's would keep a briefly-cooked egg near
 * boiling in open air, which is badly wrong for short cooks.
 *
 * `tauAirScale` is a calibration multiplier on that decay - this is the
 * least-verified part of the model, so it is learned rather than asserted.
 */
export function coolingTemperature(
  cooling: Cooling, elapsedSincePull_s: number,
  waterAtPull_C: number, meanAtPull_C: number, tauAirScale: number,
): number {
  let target: number;
  if (cooling === 'ice') {
    target = T_ICE_BATH_C;
  } else if (cooling === 'tap') {
    target = T_COLD_TAP_C;
  } else {
    const tau = TAU_AIR * tauAirScale;
    target = T_ROOM_C + (meanAtPull_C - T_ROOM_C) * Math.exp(-elapsedSincePull_s / tau);
  }
  // Blend out of the water temperature rather than jumping, so the surface is
  // continuous at the moment of pulling. See TAU_PLUNGE.
  return target + (waterAtPull_C - target) * Math.exp(-elapsedSincePull_s / TAU_PLUNGE);
}

/**
 * The full schedule: surface temperature at time t_s, where t = 0 is when the
 * egg enters the pan and `cookEnd_s` is when it comes out.
 */
export function surfaceTemperature(
  setup: CookSetup, t_s: number, cookEnd_s: number, tauAirScale: number,
  meanAtPull_C: number, waterAtPull_C: number,
): number {
  if (t_s >= cookEnd_s) {
    return coolingTemperature(
      setup.cooling, t_s - cookEnd_s, waterAtPull_C, meanAtPull_C, tauAirScale,
    );
  }
  const standing = setup.afterBoil === 'off';
  if (setup.startMode === 'cold') {
    if (t_s < setup.timeToBoil_s) {
      return rampTemperature(t_s, setup.timeToBoil_s, setup.ambient_C, setup.boiling_C);
    }
    if (!standing) return setup.boiling_C;
    return standingTemperature(
      t_s - setup.timeToBoil_s, setup.boiling_C, setup.ambient_C, setup.timeToBoil_s,
    );
  }
  // Hot start: the water is already boiling but dips when the eggs go in.
  const dip = dipMagnitude(setup);
  // With the burner on it pulls the dip back over roughly a minute. With the
  // burner off nothing pulls it back: the dip is permanent, and the water falls
  // from there. This is why the standing method is so much more sensitive to
  // pan size than a boiling one - the same eggs take a bite out of the only
  // heat left in the room.
  if (!standing) return setup.boiling_C - dip * Math.exp(-t_s / TAU_DIP_RECOVERY);
  return standingTemperature(t_s, setup.boiling_C - dip, setup.ambient_C, setup.timeToBoil_s);
}

/** Water temperature at the moment the egg goes in - the sphere's initial
 *  surface condition. */
export function initialSurfaceTemperature(setup: CookSetup): number {
  return surfaceTemperature(setup, 0.0, Number.POSITIVE_INFINITY, 1.0, 0.0, 0.0);
}

File: ios/EggTimerCore/Sources/EggTimerCore/Protocol.swift
import Foundation

/// The water temperature schedule the egg's surface actually sees, across every
/// phase of the cook: pan ramp, boil, and cooling.
/// Transliterated from `src/core/protocol.ts`.

/// Cold start: eggs go in the cold pan and heat with the water. Hot start:
/// eggs are lowered into water already boiling.
public enum StartMode: String, Sendable, Codable {
    case cold, hot
}

/// What happens after the egg comes out. Not a detail: it changes the peak
/// yolk temperature by ~20 C.
public enum Cooling: String, Sendable, Codable {
    case ice, tap, counter
}

/// What happens to the burner once the water boils. `hold` is what every recipe
/// silently assumes; `off` is the standing method.
public enum HeatAfterBoil: String, Sendable, Codable {
    case hold, off
}

public struct CookSetup: Sendable, Codable, Equatable {
    public var startMode: StartMode
    /// Egg temperature when it goes in, C.
    public var eggStartC: Double
    /// Room air temperature, C. The pan starts here on a cold start, the water
    /// decays toward it with the heat off, and a rested egg cools toward it.
    public var ambientC: Double
    /// Boiling point at the user's altitude, C.
    public var boilingC: Double
    /// Measured time to a full rolling boil, s. On a cold start it is how long
    /// the ramp lasts. With the heat off it also sets the pan's loss time
    /// constant - it is the only measurement of the pan there is.
    public var timeToBoilS: Double
    public var cooling: Cooling
    /// Water volume, litres.
    public var waterLitres: Double
    /// Burner after the boil. The TypeScript leaves this optional and treats
    /// absent as `hold`; here the default does the same job.
    public var afterBoil: HeatAfterBoil
    public var eggCount: Double

    /* The egg's MASS is not here. It was, and it duplicated `Egg.massKg` - so
     * `simulate(egg, setup, ...)` took the same number twice and every caller
     * had to keep the two in step by hand. The setup is the POT; the egg is the
     * egg. */

    public init(
        startMode: StartMode, eggStartC: Double, ambientC: Double, boilingC: Double,
        timeToBoilS: Double, cooling: Cooling, waterLitres: Double,
        afterBoil: HeatAfterBoil = .hold, eggCount: Double
    ) {
        self.startMode = startMode
        self.eggStartC = eggStartC
        self.ambientC = ambientC
        self.boilingC = boilingC
        self.timeToBoilS = timeToBoilS
        self.cooling = cooling
        self.waterLitres = waterLitres
        self.afterBoil = afterBoil
        self.eggCount = eggCount
    }
}

public enum Protocols {
    /// Pan heating ramp: constant power into a lumped water mass with Newtonian
    /// losses, clamped by boiling. One measurement (time to boil) plus RAMP_R
    /// determines the whole curve.
    public static func rampTemperature(
        tS: Double, timeToBoilS: Double, ambientC: Double, boilingC: Double
    ) -> Double {
        if tS >= timeToBoilS { return boilingC }
        let tau = timeToBoilS / log(Constants.rampR / (Constants.rampR - 1.0))
        let t = ambientC + Constants.rampR * (boilingC - ambientC) * (1.0 - exp(-tS / tau))
        return t > boilingC ? boilingC : t
    }

    /// How far the water drops when cold eggs go into boiling water, C.
    /// Straight energy balance over water + eggs.
    public static func dipMagnitude(_ egg: Egg, _ setup: CookSetup) -> Double {
        let waterCapacity = setup.waterLitres * Constants.cWater
        let eggCapacity = setup.eggCount * egg.massKg * Constants.cEgg
        let total = waterCapacity + eggCapacity
        if total <= 0.0 { return 0.0 }
        return eggCapacity * (setup.boilingC - setup.eggStartC) / total
    }

    /// The pan's Newtonian loss time constant, s - the same quantity that
    /// shapes the ramp, so the measured time to boil already contains it.
    public static func panTimeConstant(_ timeToBoilS: Double) -> Double {
        Constants.tauStandingScale * timeToBoilS / log(Constants.rampR / (Constants.rampR - 1.0))
    }

    /// Water temperature once the heat is off: Newtonian cooling of the pan
    /// toward the room, starting from `fromC`.
    public static func standingTemperature(
        elapsedSinceOffS: Double, fromC: Double, ambientC: Double, timeToBoilS: Double
    ) -> Double {
        let tau = panTimeConstant(timeToBoilS)
        if !(tau > 0.0) { return fromC }
        return ambientC + (fromC - ambientC) * exp(-elapsedSinceOffS / tau)
    }

    /// Surface temperature during cooling. Ice and tap are effectively
    /// Dirichlet; still air is not, so the surface tracks the egg's own bulk
    /// temperature decaying toward the room.
    public static func coolingTemperature(
        setup: CookSetup, elapsedSincePullS: Double,
        waterAtPullC: Double, meanAtPullC: Double, tauAirScale: Double
    ) -> Double {
        var target: Double
        switch setup.cooling {
        case .ice:
            target = Constants.tIceBathC
        case .tap:
            // Mains water is not room temperature: nearer ground temperature.
            target = Constants.tColdTapC
        case .counter:
            let tau = Constants.tauAir * tauAirScale
            target = setup.ambientC
                + (meanAtPullC - setup.ambientC) * exp(-elapsedSincePullS / tau)
        }
        // Blend out of the water temperature rather than jumping, so the
        // surface is continuous at the moment of pulling.
        return target + (waterAtPullC - target) * exp(-elapsedSincePullS / Constants.tauPlunge)
    }

    /// Water temperature while the egg is still in the pan, `tS` after it went
    /// in: ramp, boil, dip, and with the heat off the pan cooling toward the
    /// room.
    public static func bathTemperature(_ egg: Egg, _ setup: CookSetup, tS: Double) -> Double {
        let standing = setup.afterBoil == .off
        if setup.startMode == .cold {
            if tS < setup.timeToBoilS {
                return rampTemperature(
                    tS: tS, timeToBoilS: setup.timeToBoilS,
                    ambientC: setup.ambientC, boilingC: setup.boilingC
                )
            }
            if !standing { return setup.boilingC }
            return standingTemperature(
                elapsedSinceOffS: tS - setup.timeToBoilS, fromC: setup.boilingC,
                ambientC: setup.ambientC, timeToBoilS: setup.timeToBoilS
            )
        }
        // Hot start: already boiling, but it dips when the eggs go in.
        let dip = dipMagnitude(egg, setup)
        // With the burner on it pulls the dip back over roughly a minute. With
        // the burner off nothing pulls it back: the dip is permanent.
        if !standing {
            return setup.boilingC - dip * exp(-tS / Constants.tauDipRecovery)
        }
        return standingTemperature(
            elapsedSinceOffS: tS, fromC: setup.boilingC - dip,
            ambientC: setup.ambientC, timeToBoilS: setup.timeToBoilS
        )
    }

    /// Water temperature at the moment the egg goes in.
    public static func initialSurfaceTemperature(_ egg: Egg, _ setup: CookSetup) -> Double {
        bathTemperature(egg, setup, tS: 0.0)
    }
}
